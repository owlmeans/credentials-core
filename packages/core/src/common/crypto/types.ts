/**
 *  Copyright 2023 OwlMeans, Inc
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import type { encodeBase58, decodeBase58, toBeArray, getBytes } from 'ethers/lib.esm/utils'
import type { sha256, randomBytes } from 'ethers/lib.esm/crypto'
import type { HDNodeWallet } from 'ethers/lib.esm/wallet'
import type { secp256k1 } from '@noble/curves/secp256k1'

export type CryptoHelper = {
  buildSignSuite: (keyOptions: BuildSignSignatureOptions) => Object

  hash: (data: Buffer | string) => string

  hashBytes: (data: Buffer) => string

  sign: (data: string | Uint8Array, key: string) => string

  getRandomBytes: (size: number) => Uint8Array

  verify: (signature: string, data: string | Uint8Array, key: string) => boolean

  normalizePassword: (password: string) => Uint8Array

  encrypt: (body: string, password: string) => Promise<string>

  decrypt: (chiper: string, password: string) => Promise<string>

  makeDerivationPath: (index?: number, change?: number, account?: number, bc?: string) => string

  base58: () => Base58Lib

  makeId: (key: string, payload?: string, expand?: boolean) => string

  /**
   * @bug seed param implicitly requires Buffer not Uint8Array!
   */
  getKey: (seed: Uint8Array, derivationPath?: string) => CryptoKey & { dp: string }
}

export interface CryptoAdapter {
  base58: {
    encode: typeof encodeBase58
    decode: typeof decodeBase58
    toArray: typeof toBeArray
  }

  sha256: {
    toBytes: typeof getBytes
    hash: typeof sha256
  }

  aes: {
    encoder: any
  }

  secp: {
    sign: typeof secp256k1.sign,
    verify: typeof secp256k1.verify
  },

  random: typeof randomBytes

  WalletClass: typeof HDNodeWallet

  _wallets: {
    [key: string]: HDNodeWallet
  }

  setBase58Impl: (encode: typeof encodeBase58, decode: typeof decodeBase58, toArray: typeof toBeArray) => void
  setSha256Impl: (hash: typeof sha256, toBytes: typeof getBytes) => void
  setAesImpl: (encoder: any) => void
  setRandomImpl: (random: typeof randomBytes) => void
  setSecpImpl: (sign: typeof secp256k1.sign, _verify: typeof secp256k1.verify) => void
}

export type Base58Lib = {
  encode: (a: Uint8Array) => string
  decode: (a: string) => Uint8Array
}

export type BuildSignSignatureOptions = {
  publicKey: string
  privateKey: string
  id: string
  controller: string
}

export type CryptoKey = {
  id?: string
  pk?: string
  pubKey?: string
  nextKeyDigest?: string,
  fragment?: string
}

export const COMMON_CRYPTO_ERROR_NOPK = 'COMMON_CRYPTO_ERROR_NOPK'
export const COMMON_CRYPTO_ERROR_NOPUBKEY = 'COMMON_CRYPTO_ERROR_NOPUBKEY'
export const COMMON_CRYPTO_ERROR_NOID = 'COMMON_CRYPTO_ERROR_NOID'
export const COMMON_CRYPTO_ERROR_ISNOTFULL = 'COMMON_CRYPTO_ERROR_ISNOTFULL'
export const COMMON_CRYPTO_ERROR_NOKEY = 'COMMON_CRYPTO_ERROR_NOKEY'