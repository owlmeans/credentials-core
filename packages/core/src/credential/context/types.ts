import { CryptoHelper } from "@owlmeans/regov-ssi-common"
import { KeyChainWrapper } from "../../keys/types"

import { DIDDocument, DIDRegistryWrapper } from '@owlmeans/regov-ssi-did'
import {
  Credential,
  CredentialContextType,
  CredentialSubject,
  WrappedDocument,
  CredentialType,
  Presentation,
  PresentationHolder,
  UnsignedCredential,
  UnsignedPresentation,
  ContextSchema,
  Validated
} from "../types"


export type BuildCommonContextMethod = (options: {
  keys: KeyChainWrapper
  crypto: CryptoHelper
  did: DIDRegistryWrapper
  defaultSchema?: string
}) => Promise<CommonContext>

export type CommonContext = {
  keys: KeyChainWrapper
  crypto: CryptoHelper
  did: DIDRegistryWrapper
  buildLDContext: (url: string, extendedCtx?: ContextSchema, baseUrl?: string) => ContextSchema
  buildCredential: BuildCredentialMethod
  signCredential: SignCredentialMethod
  verifyCredential: VerfiyCredentailMethod
  buildPresentation: BuildPresentationMethod
  signPresentation: SignPresentationMethod
  verifyPresentation: VerifyPresentationMethod
}

export type BuildCredentialMethod = <
  SubjectType extends WrappedDocument = WrappedDocument,
  Subject extends CredentialSubject<SubjectType> = CredentialSubject<SubjectType>,
  Unsigned extends UnsignedCredential<Subject> = UnsignedCredential<Subject>
  >(options: BuildCredentailOptions<SubjectType>) => Promise<Unsigned>

// export type CommonContextType = TContext

export type BuildCredentailOptions<
  SubjectType extends WrappedDocument = WrappedDocument,
  Subject extends CredentialSubject<SubjectType> = CredentialSubject<SubjectType>
  > = {
    id: string,
    type: CredentialType
    holder: string,
    subject: Subject,
    issueanceDate?: string
    context: CredentialContextType
  }

export type SignCredentialMethod = <
  Subject extends CredentialSubject = CredentialSubject,
  CredentialT extends Credential<Subject> = Credential<Subject>,
  CredentialU extends UnsignedCredential<Subject> = UnsignedCredential<Subject>
  >(
  unsingedCredential: CredentialU,
  issuer: DIDDocument,
  options?: SignCredentialOptions
) =>
  Promise<CredentialT>

export type SignCredentialOptions = {
  buildProofPurposeOptions?: () => Promise<Object>
  keyId?: string
}

export type VerfiyCredentailMethod = (
  credential: Credential,
  did?: DIDDocument | string,
  keyId?: string
) => Promise<[boolean, VerificationResult]>

export type VerificationResult<CredentialT extends Credential = Credential>
  = Validated<CredentialT>

export type BuildPresentationMethod = <
  CredentialT extends Credential = Credential,
  Holder extends PresentationHolder = PresentationHolder
  >(
  credentials: CredentialT[],
  options: BuildPresentationOptions
) => Promise<UnsignedPresentation<CredentialT, Holder>>

export type BuildPresentationOptions = {
  id?: string
  type?: string | string[]
  context?: CredentialContextType
  holder: string
}

export type SignPresentationMethod = <
  CredentialT extends Credential = Credential,
  Holder extends PresentationHolder = PresentationHolder
  >(
  unsignedPresentation: UnsignedPresentation<CredentialT, Holder>,
  holder: DIDDocument,
  options?: SignPresentationOptions
) => Promise<Presentation<CredentialT, Holder>>

export type SignPresentationOptions = {
  buildProofPurposeOptions?: () => Promise<Object>
  challange?: string
  domain?: string
  keyId?: string
}

export type VerifyPresentationMethod = (
  presentation: Presentation,
  didDoc?: DIDDocument
) => Promise<[boolean, VerifyPresentationResult]>

export type VerifyPresentationResult<PresentationT extends Presentation = Presentation>
  = Validated<PresentationT>

