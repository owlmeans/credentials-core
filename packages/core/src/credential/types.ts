import {
  MaybeArray as AffinidiMaybeArray,
  SimpleThing,
  TContext,
  VCV1,
  VCV1Subject,
  VCV1Type,
  VCV1Unsigned,
  VPV1,
  VPV1Holder,
  VPV1Type,
  VPV1Unsigned,
  ContextObj
} from "@affinidi/vc-common";

import { Validatied as AffinidiValidatied } from "@affinidi/vc-common/dist/verifier/util"

export type Validated<T> = AffinidiValidatied<T>

export type MaybeArray<T> = AffinidiMaybeArray<T>

export type ContextSchema = ContextObj

export type Credential<Subject extends MaybeArray<CredentialSubject> = MaybeArray<CredentialSubject>>
  = VCV1<Subject>

export type Identity<Subject extends MaybeArray<IdentitySubject> = MaybeArray<IdentitySubject>>
  = Credential<Subject>

export type IdentitySubject<
  Type extends WrappedDocument = WrappedDocument,
  ExtendedType extends {} = {}
  > = CredentialSubject<Type, ExtendedType>

export type CredentialSubjectProperty<Type extends CredentialSubject = CredentialSubject>
  = Type

export type CredentialSubject<
  SubjectType extends WrappedDocument = WrappedDocument,
  ExtendedType extends {} = {}
  > = VCV1Subject<SubjectType> & ExtendedType

export type WrappedDocument<ExtendedData extends {} = {}> = SimpleThing & ExtendedData

export type CredentialContextType = TContext

export type UnsignedCredential<
  Subject extends MaybeArray<CredentialSubject> = MaybeArray<CredentialSubject>
  > = VCV1Unsigned<Subject>

export type CredentialType = VCV1Type

export type UnsignedPresentation<
  CredentialT extends Credential = Credential,
  Holder extends PresentationHolder = PresentationHolder,
  > = VPV1Unsigned<CredentialT, PresentationType, Holder>


export type Presentation<
  CredentialT extends Credential = Credential,
  Holder extends PresentationHolder = PresentationHolder
  > = VPV1<CredentialT, PresentationType, Holder>

export type PresentationHolder = VPV1Holder

export type PresentationType = VPV1Type

export const BASE_CREDENTIAL_TYPE = 'VerifiableCredential'
export const BASE_PRESENTATION_TYPE = 'VerifiablePresentation'

export const ERROR_INVALID_PRESENTATION = 'ERROR_INVALID_PRESENTATION'