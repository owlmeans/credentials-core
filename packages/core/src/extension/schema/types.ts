/**
 *  Copyright 2023 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { MaybeArray } from "../../common"
import { BasicCredentialType, CredentialSchema, MultiSchema, Credential, Presentation, UnsignedCredential } from "../../vc"
import { CredentialWrapper, WalletWrapper } from '../../wallet'
import { Extension } from "../ext"
import { ExtensionRegistry } from "../registry"


export type ExtensionSchema = {
  details: ExtensionDetails
  credentials?: { [key: string]: CredentialDescription }
  events?: ExtensionEvent[]
}

export type ExtensionDetails = {
  name: string
  code: string
  defaultCredType?: string
  types?: ExtensionTypes
  organization?: string
  home?: string
  schemaBaseUrl?: string
}

export type ExtensionTypes = {
  claim?: string
  offer?: string
}

export type CredentialDescription<
  Subject extends {} = {},
  Schema extends CredentialSchema = CredentialSchema
> = {
  defaultNameKey?: string
  mainType: string
  mandatoryTypes?: BasicCredentialType
  credentialContext: MultiSchema
  contextUrl?: string
  evidence?: MaybeArray<CredentialEvidenceDesctiption>
  credentialSchema?: MaybeArray<Schema>
  registryType?: string
  claimType?: string
  offerType?: string
  refuseType?: string
  requestType?: string
  responseType?: string
  defaultSubject?: Subject
  arbitraryEvidence?: boolean
  verfiableId?: CredentialIdMeta
  sourceType?: string
  metaRole?: MetaRole
}

export type MetaRole = typeof META_ROLE_CREDENTIAL
  | typeof META_ROLE_CLAIM
  | typeof META_ROLE_OFFER
  | typeof META_ROLE_REQUEST
  | typeof META_ROLE_RESPONSE

export const META_ROLE_CREDENTIAL = 'credential'
export const META_ROLE_CLAIM = 'claim'
export const META_ROLE_OFFER = 'offer'
export const META_ROLE_REQUEST = 'request'
export const META_ROLE_RESPONSE = 'response'

export type CredentialIdMeta = {
  fields: string[]
}

export type CredentialEvidenceDesctiption = {
  type: string
  signing?: boolean
  schema?: MaybeArray<CredentialSchema>
}

export type ExtensionEvent = {
  trigger: MaybeArray<string>
  code?: string
  filter?: ExtensionEventFilter
  method?: EventObserverMethod
}

export type EventObserverMethod = (
  wallet: WalletWrapper,
  params: EventParams
) => Promise<boolean | undefined | void>

export type EventParams = { ext?: Extension }

export type ExtensionEventFilter =
  (wallet: WalletWrapper, params: EventParams) => Promise<boolean>

export const EXTENSION_TRIGGER_AUTHENTICATION = 'wallet:authentication'
export const EXTENSION_TRIGGER_AUTHENTICATED = 'wallet:authenticated'
export const EXTENSION_TRIGGER_UNAUTHENTICATED = 'wallet:unauthenticated'
export const EXTENSION_TRIGGER_INIT_SENSETIVE = 'wallet:init-sensetive'
export const EXTENSION_TRIGGER_ADD_CREDENTIAL = 'wallet:add-credential'
export const EXTENSION_TRIGGER_REMOVE_CREDENTIAL = 'wallet:remove-credential'
export const EXTENSION_TRIGGER_DEFAULT_SIGNATURE = 'signer:default-signature'
export const EXTENSION_TRIGGER_INCOMMING_DOC_RECEIVED = 'documnet:received'
export const EXTENSION_TRIGGER_RETRIEVE_NAME = 'credentail:get-name'
export const EXTENSION_TRIGGER_PRODUCE_IDENTITY = 'identity:produce'

export type CredentialEventParams = EventParams & {
  item: CredentialWrapper
}

export type IncommigDocumentEventParams = EventParams & {
  credential: Credential | Presentation | UnsignedCredential
  statusHandler: {
    successful: boolean
  }
  cleanUp: () => void
}

export type RetreiveNameEventParams = EventParams & {
  credential: Credential
  setName: (name: string) => void
}

export type InitSensetiveEventParams = EventParams & {
  extensions: ExtensionRegistry
}