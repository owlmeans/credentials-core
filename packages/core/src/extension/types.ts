/**
 *  Copyright 2024 OwlMeans, Inc
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
import { Extension, ExtensionServiceBuilder } from './ext/types'
import { CredentialDescription, ExtensionDetails } from './schema'

export interface ExtensionBuilder {
  ns?: string
  ext?: Extension
  translations: Record<string, Record<string, unknown>>
  details: ExtensionDetails
  credentials: { [key: string]: CredentialDescription }
  factoryBuilder?: ExtensionServiceBuilder

  setNamespace: (ns: string) => void
  addTranslation: (lng: string, resource: Record<string, unknown>) => void
  addCredential: (description: Partial<CredentialDescription>, subject?: Record<string, string | null>) => CredentialDescription
  build: () => Extension
}

export interface ExtensionBuilderParams {

}
