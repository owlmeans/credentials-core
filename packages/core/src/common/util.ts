/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { MaybeArray, Idish } from "./types"

export const normalizeValue = <Type>(value: MaybeArray<Type>): Type[] => {
  return Array.isArray(value)
    ? value
    : value ? [value] : []
}

export const simplifyValue = <Type>(value: MaybeArray<Type>) => {
  return Array.isArray(value) && value.length < 2
    ? value.length === 0 ? undefined : value[0]
    : value
}

export const singleValue = <Type>(value: MaybeArray<Type>): Type | undefined => {
  return Array.isArray(value)
    ? value.length > 0 ? value[0] : undefined
    : value
}

export const mapValue = async <Type, Res>(
  value: MaybeArray<Type>,
  callback: (value: Type) => Promise<Res>
): Promise<Res[]> => {
  return Promise.all(normalizeValue(value).map(callback))
}

export const addToValue = <Type>(value: MaybeArray<Type> | undefined, add: MaybeArray<Type>): MaybeArray<Type> => {
  if (value) {
    return [...normalizeValue(value), ...normalizeValue(add)]
  }

  return add
}

export const extractId = (id: Idish) => typeof id === 'string' ? id : id.id

export const getDeepValue = <Type, Subject extends {}>(subject: Subject, key: string, def?: Type): Type => {
  const parts = key.split('.')
  return parts.reduce((value, key) => {
    if (subject.hasOwnProperty(key)) {
      const tmp = (subject as any)[key]
      subject = tmp

      return tmp
    }

    return def
  }, def) as Type
}