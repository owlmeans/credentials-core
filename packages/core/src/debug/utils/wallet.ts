import { nodeCryptoHelper } from "@owlmeans/regov-ssi-common"
import { DIDDocument } from "@owlmeans/regov-ssi-did"
import {
  buildWalletWrapper,
  Identity,
  WalletWrapper,
  identityHelper,
  IdentitySubject,
  WrappedDocument
} from "../../index"

export namespace TestUtil {
  export type IdentityFields = {
    firstname: string
    lastname: string
  }

  export const IDENTITY_TYPE = 'TestUtilIdentity'

  export class Wallet {

    constructor(public wallet: WalletWrapper, public name: string) {
    }

    static async setup(name: string) {
      const walletUtil = new Wallet(await buildWalletWrapper(nodeCryptoHelper, '11111111',
        { name, alias: name },
        {
          prefix: process.env.DID_PREFIX,
          defaultSchema: 'https://owlmeans.com'
        }
      ), name)

      return walletUtil
    }

    async produceIdentity() {
      return await identityHelper<IdentityFields>(
        this.wallet,
        this.wallet.ctx.buildLDContext('identity', {
          xsd: 'http://www.w3.org/2001/XMLSchema#',
          firstname: { '@id': 'scm:firstname', '@type': 'xsd:string' },
          lastname: { '@id': 'scm:lastname', '@type': 'xsd:string' }
        })
      ).createIdentity(
        IDENTITY_TYPE,
        {
          firstname: this.name,
          lastname: 'Lastname'
        },
      )
    }

    async claimDocument() {
      return 
    }
  }
}