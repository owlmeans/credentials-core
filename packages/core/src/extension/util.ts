/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { normalizeValue } from "../common"
import { BasicCredentialType } from "../vc"
import {
  CredentialService, CredentialServiceBuilder, defaultBuildMethod, defaultClaimMethod,
  defaultOfferMethod, defaultRequestMethod, defaultRespondMethod, defaultSignMethod, defaultValidateMethod,
  Extension
} from "./ext"
import { CredentialDescription } from "./schema"


export const findAppropriateCredentialType = (
  ext: Extension, types: BasicCredentialType, defaultType: string
): string => {
  types = normalizeValue(types)

  return (Object.entries(ext.factories).map(([type]) => type)
    .find(type => types.includes(type)) || defaultType)
}

const methodMap = {
  'produceBuildMethod': 'build',
  'produceSignMethod': 'sign',
  'produceValidateMethod': 'validate',
  'produceClaimMethod': 'claim',
  'produceOfferMethod': 'offer',
  'produceRequestMethod': 'request',
  'produceRespondMethod': 'respond'
}

export const addFactoriesToExt = (ext: Extension, type: string, factories: CredentialServiceBuilder) => {
  const description = (ext.schema.credentials as { [key: string]: CredentialDescription })[type]
  ext.factories[type] = {
    ...{
      build: defaultBuildMethod(description),
      sign: defaultSignMethod(description),
      validate: defaultValidateMethod(description),
      claim: defaultClaimMethod(description),
      offer: defaultOfferMethod(description),
      request: defaultRequestMethod(description),
      respond: defaultRespondMethod(description)
    },
    ...Object.entries(factories).reduce((_facts, [method, builder]) => {
      if (ext.schema.credentials) {
        return {
          ..._facts,
          [(methodMap as any)[method]]: builder(description)
        }
      }
      return _facts
    }, {} as CredentialService)
  }
}