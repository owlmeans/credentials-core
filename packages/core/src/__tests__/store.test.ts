import { nodeCryptoHelper } from "@owlmeans/regov-ssi-common"
import { EncryptedStore } from "../store"
import { WalletWrapper } from "../wallet/types"
import { buildWalletWrapper } from "../wallet/wallet"

import util from 'util'
import { KEYCHAIN_DEFAULT_KEY } from "../keys/types"
util.inspect.defaultOptions.depth = 6


describe('Wallet', () => {

  const _ctx: {
    seed: string
    wallet?: WalletWrapper
    store?: EncryptedStore
  } = {
    seed: 'CGETek79CLUkP9vB48Tnj7VEqiWTmD46RScExYGieiE2'
  }

  it('creates store', async () => {
    const wallet = _ctx.wallet = await buildWalletWrapper(nodeCryptoHelper, '11111111', {
      alias: 'current',
      name: 'Some name'
    }, { key: { seed: _ctx.seed } })

    expect(wallet.store.alias).toBe('current')
  })

  it('exports store', async () => {
    if (!_ctx.wallet) {
      throw new Error('No wallet from pervious test')
    }

    const store = _ctx.store = await _ctx.wallet.export('11111111')

    expect(store.alias).toBe('current')
  })

  it('imports store back', async () => {
    if (!_ctx.store) {
      throw new Error('No store from pervious test')
    }

    const newWallet = await buildWalletWrapper(nodeCryptoHelper, '11111111', _ctx.store)

    expect(newWallet.wallet).toMatchSnapshot({
      keyChain: {
        keys: {
          [KEYCHAIN_DEFAULT_KEY]: {
            seed: expect.any(String),
            rotations: [
              { private: expect.any(String) },
              {
                private: expect.any(String),
                public: expect.any(String)
              },
            ]
          }
        }
      }
    })
  })
})