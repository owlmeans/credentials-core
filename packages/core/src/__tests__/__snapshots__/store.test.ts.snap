// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet imports store back 1`] = `
Object {
  "keyChain": Object {
    "defaultKey": "_identity",
    "keys": Object {
      "_identity": Object {
        "alias": "_identity",
        "currentRotation": 0,
        "dp": Array [
          0,
        ],
        "id": "8MbyK1XFhCf6tWCEuqrR5XGXAho6FhdVQm9awBRYbKzj",
        "rotations": Array [
          Object {
            "digest": "8MbyK1XFhCf6tWCEuqrR5XGXAho6FhdVQm9awBRYbKzj",
            "dp": Array [
              0,
            ],
            "future": false,
            "nextDigest": "c2xIrRxBLwPDlHhd624HwwkmYT0UNzSfVL5AuYQHK/N3NFcZ5SqBZ7HvutGiZ5toqYaOcTPTr8/WwqvvIIBcfg==",
            "opened": false,
            "private": Any<String>,
            "public": "grUKfJPKYcxsrr3Kz2DAZEWCELAADMigL3eqdqP1Za6q",
            "safe": false,
            "type": "BIP32",
          },
          Object {
            "digest": "4CqYJPSrK2tiUYm8sUJsnUHN2qsGJ3r1Xsq6LY6baNvG",
            "dp": Array [
              1,
            ],
            "future": true,
            "opened": false,
            "private": Any<String>,
            "public": Any<String>,
            "safe": false,
            "type": "BIP32",
          },
        ],
        "safe": false,
        "seed": Any<String>,
        "type": "BIP32",
      },
    },
  },
  "registry": Object {
    "peer": Object {
      "dids": Array [],
    },
    "personal": Object {
      "dids": Array [],
    },
  },
}
`;
